### PR TITLE
Upload directory and multiple files, show upload progress bar, and keyboard-control upload menu

### PR DESCRIPTION
**Pull request by Levilian**
Implement #1050

### CODE REVIEW

**Comment by percyliang** [comment]
Seems like we have a lot of files added unintentionally.  Probably best to recreate the branch to not have these in our history.

**Comment by yipenghe** [comment]
Lol there still several unwanted files (#codalab_service.py# ...etc.

**Comment by yipenghe** [comment]
There seems to be some very big changes, we should test it thoroughly on dev before merging to master

**Comment by percyliang** [comment]
Are we sure all the original extraneous files are actually purged from the history?  I'd feel safer having a new branch...

**Comment by Levilian** [comment]
> Are we sure all the original extraneous files are actually purged from the history? I'd feel safer having a new branch...

Yup! I rebased the branch and squashed the commits

File: codalab/lib/upload_manager.py [hunk 1]
```diff
@@ -45,77 +46,100 @@ def upload_to_bundle_store(
         |simplify_archives|: whether to simplify unpacked archives so that if they
                              contain a single file, the final path is just that file,
                              not a directory containing that file.
-
+        |is_dir_upload|: whether the incoming sources are individual files that need to be rebuild
+        into a directory or merely individual files.
```
Comment (percyliang) [inline codalab/lib/upload_manager.py]: Say more about what `sources` are in this case.

File: codalab/lib/upload_manager.py [hunk 2]
```diff
@@ -45,77 +46,100 @@ def upload_to_bundle_store(
...
-                is_url, is_local_path, is_fileobj, filename = self._interpret_source(source)
-                source_output_path = os.path.join(bundle_path, filename)
-                if is_url:
-                    if git:
-                        source_output_path = file_util.strip_git_ext(source_output_path)
-                        file_util.git_clone(source, source_output_path)
-                    else:
-                        file_util.download_url(source, source_output_path)
-                        if unpack and self._can_unpack_file(source_output_path):
+        if is_dir_upload:
```
Comment (percyliang) [inline codalab/lib/upload_manager.py]: I thought we went with the version where we zip up things on the client side, but we switched back to uploading individual files?

File: frontend/src/components/worksheets/NewUpload/NewUpload.jsx [hunk 3]
```diff
@@ -41,46 +42,163 @@ class NewUpload extends React.Component<{
...
-        if (!file) {
+    uploadFile = (files) => {
+        if (!files) {
+            return;
+        }
+        let folderSize = 0;
+        for (const file of files) {
+            folderSize += file.size;
+        }
+        if (folderSize/1024/1024/1024 > 2) {
```
Comment (percyliang) [inline frontend/src/components/worksheets/NewUpload/NewUpload.jsx]: Prefer multiplication on the right-hand side.

File: frontend/src/components/worksheets/NewUpload/NewUpload.jsx [hunk 4]
```diff
@@ -41,46 +42,163 @@ class NewUpload extends React.Component<{
...
+            if (after_sort_key) {
+                url += `&after_sort_key=${ after_sort_key }`;
+            }
+            $.ajax({
+                url,
+                data: JSON.stringify(createBundleData),
+                contentType: 'application/json',
+                type: 'POST',
+                success: (data, status, jqXHR) => {
+                    var bundleUuid = data.data[0].id;
```
Comment (percyliang) [inline frontend/src/components/worksheets/NewUpload/NewUpload.jsx]: Use `let` or better yet `const`

File: frontend/src/components/worksheets/items/ActionButtons.js [hunk 5]
```diff
@@ -44,18 +59,43 @@ class ActionButtons extends React.Component<{
...
+                            aria-label='Add New Upload'
+                            onClick={this.handleClick}
+                            disabled={!editPermission}
+                        >
                             <UploadIcon className={classes.buttonIcon} />
                             Upload
-                        </label>
-                    </Button>
+                        </Button>
+                        <Menu
```
Comment (yipenghe) [inline frontend/src/components/worksheets/items/ActionButtons.js:85]: It might be better to be a dropdown menu instead of covering upload

**Review by yipenghe — COMMENTED** [review]
Using click works great!

Keyboard shortcut  (:0)doesn't seem to work, "a u" will trigger the menu to show, but "up" "down" keyboards will move both the focus on worksheet and the upload menu. Hitting "enter" also doesn't open the upload menu, but will trigger bundle detail / text editing in the worksheet area

File: frontend/src/components/worksheets/NewUpload/NewUpload.jsx [hunk 6]
```diff
@@ -41,46 +42,163 @@ class NewUpload extends React.Component<{
...
-            contentType: 'application/json',
-            type: 'POST',
-            success: (data, status, jqXHR) => {
-                var bundleUuid = data.data[0].id;
-                var reader = new FileReader();
-                reader.onload = () => {
-                    var arrayBuffer = reader.result,
-                        bytesArray = new Uint8Array(arrayBuffer);
+
+        var zip = new JSZip();
```
Comment (yipenghe) [inline frontend/src/components/worksheets/NewUpload/NewUpload.jsx]: If we zipped here, why do we need to the is_dir_upload operations?

**Comment by yipenghe** [comment]
I also like the percentage progress!

File: frontend/package.json [hunk 7]
```diff
@@ -21,16 +21,19 @@
     "jquery-ui-bundle": "^1.12.1-migrate",
     "jquery.terminal": "^2.1.2",
     "jsonapi-datastore": "^0.4.0-beta",
+    "jszip": "^3.2.2",
     "lint-staged": "^8.2.1",
     "marked": "^0.6.2",
     "material-ui-chip-input": "^1.0.0-beta.16",
     "query-string": "^6.2.0",
     "react": "^16.8.6",
+    "react-circular-progressbar": "^2.0.3",
```
Comment (epicfaace) [inline frontend/package.json:30]: Can you update the title / description of this PR to make clear that this PR also adds the progress bar?

File: frontend/src/components/worksheets/NewUpload/NewUpload.jsx [hunk 8]
```diff
@@ -41,54 +42,171 @@ class NewUpload extends React.Component<{
...
-        if (!file) {
+    uploadFile = (files) => {
+        if (!files) {
+            return;
+        }
+        let folderSize = 0;
+        for (const file of files) {
+            folderSize += file.size;
+        }
+        if (folderSize > 2*1024*1024*1024) {
```
Comment (epicfaace) [inline frontend/src/components/worksheets/NewUpload/NewUpload.jsx]: Please make this a constant -- it could be exported from codalab-worksheets/frontend/src/constants.js.

Similarly with the "2 GB" alert.

File: frontend/src/components/worksheets/NewUpload/NewUpload.jsx [hunk 9]
```diff
@@ -41,54 +42,171 @@ class NewUpload extends React.Component<{
...
         if (!files.length) {
             return;
         }
-        this.uploadFile(files[0]);
+        this.uploadFolder(files);
     }
 
-    uploadFile = (file) => {
-        if (!file) {
+    uploadFile = (files) => {
```
Comment (epicfaace) [inline frontend/src/components/worksheets/NewUpload/NewUpload.jsx]: Can you rename this method `uploadFiles`?

File: frontend/src/components/worksheets/NewUpload/NewUpload.jsx [hunk 10]
```diff
@@ -41,54 +42,171 @@ class NewUpload extends React.Component<{
...
+        if (!files) {
+            return;
+        }
+        let folderSize = 0;
```
Comment (epicfaace) [inline frontend/src/components/worksheets/NewUpload/NewUpload.jsx]: You can instead do (looks nicer, also use const instead of let):

```
const folderSize = files.reduce((a, b) => a.size + b.size, 0);
```

File: frontend/src/components/worksheets/NewUpload/NewUpload.jsx [hunk 11]
```diff
@@ -41,54 +42,171 @@ class NewUpload extends React.Component<{
...
+            alert('File size is large than 2 GB. Please upload your file(s) through CLI.');
             return;
         }
         const { worksheetUUID, after_sort_key } = this.props;
         const { name, description } = this.state;
-        const createBundleData = getDefaultBundleMetadata(name || file.name, description);
+
+        for (const file of files) {
+            const createBundleData = getDefaultBundleMetadata(name || file.name, description);
+            this.setState({
```
Comment (epicfaace) [inline frontend/src/components/worksheets/NewUpload/NewUpload.jsx]: Right now, you're setting `uploading: true` before each file is uploaded and calling the success/failure callbacks (and `reloadWorksheet`) after each file is uploaded. Can you make it so that the former happens before the for loop (before _all_ files are uploaded) and the latter after _all_ files have been uploaded?

File: frontend/src/components/worksheets/NewUpload/NewUpload.jsx [hunk 12]
```diff
@@ -41,54 +42,171 @@ class NewUpload extends React.Component<{
...
-                    var url =
+
+        let zip = new JSZip();
+        [...files].map(file => {
+            zip.file(file.webkitRelativePath, file);
+        });
+
+        zip.generateAsync({type:"uint8array", compression: "DEFLATE"}).then((bytesArray) => {
+
+            $.ajax({
```
Comment (epicfaace) [inline frontend/src/components/worksheets/NewUpload/NewUpload.jsx:197]: Can you unify this $.ajax call and the $.ajax call in `uploadFile` into a single function called by both functions?

File: frontend/src/components/worksheets/NewUpload/NewUpload.jsx [hunk 13]
```diff
@@ -114,47 +232,62 @@ class NewUpload extends React.Component<{
...
+                />
+                { uploading && <CircularProgressbar
                         className={ classes.progress }
                         variant="determinate"
                         value={ percentComplete }
-                        size={ 80 }
+                        text={`${percentComplete}% uploaded`}
+                        styles={buildStyles({
+                            textSize: '12px',
+                            // pathColor: 'teal',
```
Comment (epicfaace) [inline frontend/src/components/worksheets/NewUpload/NewUpload.jsx]: remove comments

File: frontend/src/components/worksheets/NewUpload/NewUpload.jsx [hunk 14]
```diff
@@ -164,11 +297,12 @@ class NewUpload extends React.Component<{
 
 const styles = (theme) => ({
     progress: {
+        "z-index": 1000,
```
Comment (epicfaace) [inline frontend/src/components/worksheets/NewUpload/NewUpload.jsx]: Not sure if this would work -- you may have to do `zIndex: 1000`

File: frontend/src/components/worksheets/Worksheet/Worksheet.js [hunk 15]
```diff
@@ -470,6 +471,7 @@ class Worksheet extends React.Component {
             focusedBundleUuidList: focusedBundleUuidList,
             showNewRun: false,
             showNewText: false,
+            // uploadAnchor: null,
```
Comment (epicfaace) [inline frontend/src/components/worksheets/Worksheet/Worksheet.js]: Why comment this out?

File: frontend/src/components/worksheets/Worksheet/Worksheet.js [hunk 16]
```diff
@@ -730,7 +731,7 @@ class Worksheet extends React.Component {
                         if (this.state.focusIndex < 0) {
                             $('html, body').animate({ scrollTop: $(document).height() }, 'fast');
                         }
-                        document.querySelector('label[for=codalab-file-upload-input]').click();
+                        document.querySelector('Button[id=upload-button]').click();
```
Comment (epicfaace) [inline frontend/src/components/worksheets/Worksheet/Worksheet.js]:
```suggestion
                        document.querySelector('#upload-button').click();
```

File: frontend/src/components/worksheets/items/ActionButtons.js [hunk 17]
```diff
@@ -44,18 +70,53 @@ class ActionButtons extends React.Component<{
...
-                        disabled={!editPermission}
-                    >
-                        <label className={classes.uploadLabel} for='codalab-file-upload-input'>
+                    <span>
+                        <Button
+                            size='small'
+                            color='inherit'
+                            id='upload-button'
+                            aria-label='Add New Upload'
+                            onClick={onShowNewUpload}
```
Comment (epicfaace) [inline frontend/src/components/worksheets/items/ActionButtons.js]: By adding the button to trigger uploads instead of the label, I think there's a possibility that you're undoing the fixes done in #1945. Can you make sure that the following things still work:
- uploads still work on chrome
- when I open the upload file dialog and exit it, then click "upload file" again, the dialog still shows up
- when I upload a file X, then click "upload file" again and select the same file X, the same file gets uploaded again

File: frontend/src/components/worksheets/Worksheet/Worksheet.js [hunk 18]
```diff
@@ -1251,9 +1252,24 @@ class Worksheet extends React.Component {
         this.setState({ BulkBundleDialog: deleteWorksheetDialog });
     }
 
+    onShowNewUpload = (e) => {
```
Comment (epicfaace) [inline frontend/src/components/worksheets/Worksheet/Worksheet.js]: Can you rename this to something like `showUploadMenu` (so that it doesn't sound too much like `onShowNewRun` or `onShowNewText`, which do very different things)? (and before #1945, we used to have a variable called `showNewUpload` as well)

File: frontend/src/components/worksheets/Worksheet/Worksheet.js [hunk 19]
```diff
@@ -1251,9 +1252,24 @@ class Worksheet extends React.Component {
...
+        e.stopPropagation();
```
Comment (epicfaace) [inline frontend/src/components/worksheets/Worksheet/Worksheet.js]: Why is this needed?

File: frontend/src/components/worksheets/Worksheet/Worksheet.js [hunk 20]
```diff
@@ -1251,9 +1252,24 @@ class Worksheet extends React.Component {
...
+        this.setState({ uploadAnchor: e.currentTarget });
+        let form = document.querySelector('Menu[id=upload-menu]');
```
Comment (epicfaace) [inline frontend/src/components/worksheets/Worksheet/Worksheet.js]: You should do `#upload-menu`

File: frontend/src/components/worksheets/Worksheet/Worksheet.js [hunk 21]
```diff
@@ -1251,9 +1252,24 @@ class Worksheet extends React.Component {
...
+        Mousetrap.unbind(['up', 'k']);
+        Mousetrap.unbind(['down', 'j']);
+        Mousetrap.unbind(['enter']);
+        Mousetrap.reset();
+        Mousetrap(form).bind(['enter'], function(e) {
+            e.preventDefault();
+            // console.log(e.target.firstElementChild.htmlFor);
```
Comment (epicfaace) [inline frontend/src/components/worksheets/Worksheet/Worksheet.js]: remove comment

File: frontend/src/components/worksheets/Worksheet/Worksheet.js [hunk 22]
```diff
@@ -1251,9 +1252,24 @@ class Worksheet extends React.Component {
...
+            document.querySelector('label[for=' + e.target.firstElementChild.htmlFor + ']').click();
+        });
```
Comment (epicfaace) [inline frontend/src/components/worksheets/Worksheet/Worksheet.js:1263]: When the user opens the menu, can they press "esc" to exit out of it?

Comment (epicfaace) [inline frontend/src/components/worksheets/Worksheet/Worksheet.js:1263]: When the user opens the menu and then exits out of it, does the normal functionality for "enter" / "up" / "down" / etc. still work?

**Review by epicfaace — CHANGES_REQUESTED** [review]
Great work so far!

File: frontend/src/components/worksheets/NewUpload/NewUpload.jsx (hunk 10 above)
Comment (Levilian) [reply frontend/src/components/worksheets/NewUpload/NewUpload.jsx]: https://developer.mozilla.org/en-US/docs/Web/API/FileList
`FileList API` is actually an object, not an array.

File: frontend/src/components/worksheets/Worksheet/Worksheet.js (hunk 22 above)
Comment (Levilian) [reply frontend/src/components/worksheets/Worksheet/Worksheet.js:1263]: yes

Comment (Levilian) [reply frontend/src/components/worksheets/Worksheet/Worksheet.js:1263]: yes

File: frontend/src/util/ws_mousetrap_fork.js [hunk 23]
```diff
@@ -1066,3 +1066,35 @@
...
     }
 })(typeof window !== 'undefined' ? window : null, typeof window !== 'undefined' ? document : null);
+
+/**
+ * adds a pause and unpause method to Mousetrap
+ * this allows you to enable or disable keyboard shortcuts
+ * without having to reset Mousetrap and rebind everything
+ */
+/* global Mousetrap:true */
+(function(Mousetrap) {
```
Comment (epicfaace) [inline frontend/src/util/ws_mousetrap_fork.js:1076]: Did you write this code yourself or was it automatically added somehow?

File: frontend/src/components/worksheets/NewUpload/NewUpload.jsx [hunk 24]
```diff
@@ -114,47 +242,60 @@ class NewUpload extends React.Component<{
...
         this.setState({ percentComplete: 0, uploading: false });
     }
-    
 
     render() {
         const { classes } = this.props;
         const { percentComplete, uploading } = this.state;
 
         return (
+
```
Comment (epicfaace) [inline frontend/src/components/worksheets/NewUpload/NewUpload.jsx]: Remove empty line

File: frontend/src/components/worksheets/NewUpload/NewUpload.jsx [hunk 25]
```diff
@@ -41,54 +43,180 @@ class NewUpload extends React.Component<{
...
-        this.uploadFile(files[0]);
+        this.uploadFolder(files);
+    }
+
+    uploadFiles = (files) => {
+        if (!files) {
+            return;
+        }
+
+        if (!files) {
```
Comment (epicfaace) [inline frontend/src/components/worksheets/NewUpload/NewUpload.jsx]: Remove this if statement

File: frontend/src/components/worksheets/NewUpload/NewUpload.jsx [hunk 26]
```diff
@@ -41,54 +43,180 @@ class NewUpload extends React.Component<{
...
-            success: (data, status, jqXHR) => {
-                var bundleUuid = data.data[0].id;
-                var reader = new FileReader();
-                reader.onload = () => {
-                    var arrayBuffer = reader.result,
-                        bytesArray = new Uint8Array(arrayBuffer);
-                    var url =
+
+        let zip = new JSZip();
+        [...files].map(file => {
```
Comment (epicfaace) [inline frontend/src/components/worksheets/NewUpload/NewUpload.jsx:191]: Can you just use files instead of ...files?

**Review by epicfaace — CHANGES_REQUESTED** [review]

File: frontend/src/components/worksheets/NewUpload/NewUpload.jsx (hunk 26 above)
Comment (Levilian) [reply frontend/src/components/worksheets/NewUpload/NewUpload.jsx:191]: File API treats files as an object, so we need to flatten it.

File: frontend/src/util/ws_mousetrap_fork.js (hunk 23 above)
Comment (Levilian) [reply frontend/src/util/ws_mousetrap_fork.js:1076]: I added it so we could perform keyboard shortcuts albeit having the upload menu.

Comment (epicfaace) [reply frontend/src/util/ws_mousetrap_fork.js:1076]: @Levilian it would be good if you moved this code to a separate file. Everything else in this file is vendor code from mousetrap, right?

File: frontend/src/components/worksheets/NewUpload/NewUpload.jsx (hunk 12 above)
Comment (epicfaace) [reply frontend/src/components/worksheets/NewUpload/NewUpload.jsx:197]: @Levilian -- this is important because they basically call the same endpoint. It's better to commit non-duplicative code, if possible, than leave things like this for future coders to refactor.
